### PR TITLE
[RF] Reduce memory consumption of MemPoolForRooSets

### PR DESCRIPTION
**Pull request by guitargeek**
The MemPoolForRooSets has two problems:

 1. Scaling issues: increasing memory and CPU consumption with
    increasing number of arenas, becuase trying to allocate a new
    non-overlapping arena often fails.

 2. Amplification of memory leaks in user code:
    If one of the elements in the arena leaks, the whole arena will
    leak. Since each arena has 6000 elements, this means a leak of a
    RooArgSet will be ampilfied by a factor 6000 in the worst case.

This commit proposes a solution to the scaling/performance issues and
the leak amplification.

Explained for the example for RooArgSet, the idea is to use each arena
120 times because this is the size of RooArgSet in bytes. This can be
done while still having unique adresses for each RooArgSet:

 1. Keep a gap of 120 bytes between each RooArgSet
 2. When an address was used before, add one byte and in can be used
    again

So at the cost of allocating 2x more memory per arena, we can reuse each
arena 120 times.

Now, the good thing is that the interval of creating/deleting
RooArgSets is usually higher than 6000. So when reusing an arena
multiple times, they will eventually be filled only by leaking
RooArgSets if there is a leak. Meaning the leak amplification effect
should be gone.

This was checked with the CMS Higgs combine toy fit example from the
Higgs combine tutorial. Without this commit, there is a GB-level leak
that can be clearly observed with `top`, but with this commit the leak
becomes negligible again.

The PR is still kind of a draft PR. I will check with the bot if this solution works on all platforms and then continue improving the code.

This should be backported together with https://github.com/root-project/root/pull/7935.

This should finally fix https://github.com/root-project/root/issues/7933.

Profile of the CMS toy experiment example with a ROOT Debug build:
* [without this PR](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/combine_example_2_old)
* [with this PR](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/arena_1_new)

### CODE REVIEW

**Comment by lmoneta** [comment]
This PR looks a big improvement in fixing #7890 and #7933

Here are some  test results run on Ubuntu 20

- Test create/delete a RooArgSet,  simple example shown in #7933

    - with current mater (i.e. including  #7904 and #7935): 
       - Time to create/delete 10M RooArgSet :  125 sec
       - Increase of Memory  ~ 1 GB
    - with master + this PR: 
        - Time to create/delete 10M RooArgSet :  0.8  sec
        - total memory increase + 25 Mb)
     - using 6.15 (before including current MemPoolForRooSets (-> #2866 )
          - Time to create/delete 10M RooArgSet :  2.9 sec
          - memory increase : 0 
          
- Running RooStats toy generation + likelihood evaluation on Histfactory tutorial model using in totoal 300k toys: 
```
StandardHypoTestInvDemo("$ROOTSYS/tutorials/histfactory/results/example_combined_GaussExample_model.root","combined","ModelConfig","","obsData",0,0,true,2,0,5,100000)
```

  - master :                             Time: 344 sec  Memory : ~ 3.5 GB
   - master + this PR :             Time: 157 sec   Memory:  570 Mb
   - 6.15                                   Time: 197 sec   Memory:  240 Mb
   
   So in conclusion:

 - this PR fixes the increase time spent in allocate/deallocate RooArgSet over time, which will slow down toys analysis when number of toys increase
 - there is still some increase in memory usage, but it is much smaller than before.

**Review by lmoneta — APPROVED** [review]
Very good!
Thank you Jonas !!

Before merging, let;s see if @hageboeck has some comments

**Review by hageboeck — APPROVED** [review]
I was going to say "Measure the performance", but I see that you did. 👍 
Good idea to recycle those sets, and good to keep the double delete checks, at least in debug builds.

I don't see whether the new `#ifdef`s should stay. It's better to have less `ifdef` code, because it will rot faster than the code that's actually compiled in the CI. (Exception: NDEBUG, because there are debug builds).

If it can be helpful for future debugging, OK to leave it in. If it's only to finish the development of the Arena, take it out before putting it in master.

**Comment by guitargeek** [comment]
Hi @lmoneta and @hageboeck, thanks for the profiling and the review!

Indeed, the first commit in this PR with the `ifdefs` should be dropped. With the improvements in this PR, I hope we don't have to touch the arena anytime soon again anyway.

I will not touch the second commit thought, because it has already been carefully validated by Lorenzo.

**Comment by lmoneta** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On
